File: src/main/java/com/choru/lock/infrastructure/UserPointRepository.java
package com.choru.lock.infrastructure;

import com.choru.lock.domain.UserPoint;
import jakarta.persistence.LockModeType;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;

import java.util.Optional;

public interface UserPointRepository extends JpaRepository<UserPoint, Long> {

    Optional<UserPoint> findByUserId(Long userId);

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("select up from UserPoint up where up.userId = :userId")
    Optional<UserPoint> findByUserIdWithPessimisticLock(Long userId);
}

File: src/main/kotlin/com/choru/lock/infrastructure/UserPointRepository.kt
package com.choru.lock.infrastructure

import com.choru.lock.domain.UserPoint
import jakarta.persistence.LockModeType
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Lock
import org.springframework.data.jpa.repository.Query

interface UserPointRepository : JpaRepository<UserPoint, Long> {

    fun findByUserId(userId: Long): UserPoint?

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("select up from UserPoint up where up.userId = :userId")
    fun findByUserIdWithPessimisticLock(userId: Long): UserPoint?
}
